fl4health/clients/scaffold_client.py

```diff
@@ -69,24 +68,34 @@ def get_parameters(self, config: Config) -> NDArrays:
 
     def set_parameters(self, parameters: NDArrays, config: Config) -> None:
         """
-        Assumes that the parameters being passed contain model parameters concatenated with
-        server control variates. They are unpacked for the clients to use in training
+        Assumes that the parameters being passed contain model parameters concatenated with server control variates.
+        They are unpacked for the clients to use in training. If it's the first time the model is being initialized,
+        we assume the full model is being initialized and use the FullParameterExchanger() to set all model weights
+        Args:
+            parameters (NDArrays): Parameters have information about model state to be added to the relevant client
+                model and also the server control variates (initial or after aggregation)
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
         """
         assert self.model is not None and self.parameter_exchanger is not None
 
         server_model_state, server_control_variates = self.parameter_exchanger.unpack_parameters(parameters)
         self.server_control_variates = server_control_variates
-        self.server_model_state = server_model_state
-        self.parameter_exchanger.pull_parameters(server_model_state, self.model, config)
+
+        super().set_parameters(server_model_state, config)
+
+        # Note that these are weights the require a gradient, because they are used to compute control variates
         self.server_model_weights = [
-            model_params.cpu().detach().numpy()
+            model_params.cpu().detach().clone().numpy()
```

Comment:
If we don't call clone here and the `model_params` are already on cpu, there is a memory optimization where this numpy array will **share the underlying memory of the parameters with the original tensor**. This is bad, because it means that when the model parameters are updated in training, the values in `self.server_model_weights` are **also updated**, which we really don't want.

Reply:
Note: This does not cause the same problems if the model is on GPU, as model_params.cpu() makes a copy of the tensor on cpu, distinct from the tensor being trained on the GPU.

Reply:
Very good catch! I was not aware of this at all but will be sure to keep this in mind in the future.